File: core/src/net/natruid/jungle/systems/CombatTurnSystem.java
package net.natruid.jungle.systems;

import com.artemis.Aspect;
import com.artemis.BaseEntitySystem;
import com.artemis.ComponentMapper;
import com.artemis.EntitySubscription;
import com.artemis.utils.IntBag;
import com.badlogic.gdx.utils.Array;
import net.natruid.jungle.components.StatsComponent;
import net.natruid.jungle.components.TurnComponent;
import net.natruid.jungle.components.UnitComponent;
import net.natruid.jungle.utils.Constants;
import net.natruid.jungle.utils.Logger;
import net.natruid.jungle.utils.types.Faction;

public class CombatTurnSystem extends BaseEntitySystem {
    private final Array<Faction> factionList = new Array<>();
    private int turn;
    private int currentFactionIndex = 0;
    private Phase phase = Phase.NONE;

    private BehaviorSystem behaviorSystem;
    private FlowControlSystem flowControlSystem;
    private ComponentMapper<TurnComponent> mTurn;
    private ComponentMapper<UnitComponent> mUnit;
    private ComponentMapper<StatsComponent> mStats;

    private EntitySubscription unitSubscription;

    public CombatTurnSystem() {
        super(Aspect.all(TurnComponent.class));
    }

    @Override
    protected void initialize() {
        super.initialize();
        unitSubscription = world.getAspectSubscriptionManager().get(Aspect.all(UnitComponent.class));
    }

    public int getTurn() {
        return turn;
    }

    public void start() {
        phase = Phase.START;
    }

    public void addFaction(Faction faction) {
        if (factionList.contains(faction, true)) return;
        factionList.add(faction);
    }

    public Faction getFaction() {
        return factionList.get(currentFactionIndex);
    }

    public void removeFaction(Faction faction) {
        factionList.removeValue(faction, true);
    }

    public void reset() {
        factionList.clear();
        currentFactionIndex = 0;
        phase = Phase.NONE;
    }

    @Override
    protected void processSystem() {
        switch (phase) {
            case START:
                giveTurn(factionList.get(currentFactionIndex));
                phase = Phase.READY;
                break;
            case READY:
                if (getEntityIds().isEmpty() && behaviorSystem.isReady() && flowControlSystem.isReady()) {
                    phase = Phase.NEXT_TURN;
                }
                break;
            case NEXT_TURN:
                if (currentFactionIndex >= factionList.size - 1) {
                    turn += 1;
                    currentFactionIndex = 0;
                } else {
                    currentFactionIndex += 1;
                }

                Faction nextFaction = factionList.get(currentFactionIndex);
                Logger.debug("Turn ended. Next faction:" + nextFaction);
                giveTurn(nextFaction);
                behaviorSystem.prepare();
                phase = Phase.READY;
                break;
        }
    }

    public void giveTurn(Faction faction) {
        IntBag entities = unitSubscription.getEntities();
        int[] data = entities.getData();
        for (int i = 0; i < entities.size(); i++) {
            int unit = data[i];
            UnitComponent cUnit = mUnit.get(unit);

            if (cUnit.faction != faction) continue;

            cUnit.ap = Math.min((cUnit.ap + 4 + mStats.get(unit).getAp()), Constants.MAX_AP);
            mTurn.create(unit);
        }
    }

    public void endTurn(int unit) {
        if (!mTurn.has(unit)) return;
        mTurn.remove(unit);
    }

    private enum Phase {NONE, START, READY, NEXT_TURN}
}

File: core/src/net/natruid/jungle/systems/CombatTurnSystem.kt
package net.natruid.jungle.systems

import com.artemis.Aspect
import com.artemis.BaseEntitySystem
import com.artemis.ComponentMapper
import net.natruid.jungle.components.StatsComponent
import net.natruid.jungle.components.TurnComponent
import net.natruid.jungle.components.UnitComponent
import net.natruid.jungle.utils.Constants
import net.natruid.jungle.utils.Faction
import net.natruid.jungle.utils.Logger

class CombatTurnSystem : BaseEntitySystem(Aspect.all(TurnComponent::class.java)) {
    private enum class Phase { NONE, START, READY, NEXT_TURN }

    var turn = 0
        private set

    val faction get() = factionList[currentFactionIndex]

    private val factionList = ArrayList<Faction>()
    private var currentFactionIndex = 0
    private var phase = Phase.NONE

    private lateinit var behaviorSystem: BehaviorSystem
    private lateinit var flowControlSystem: FlowControlSystem
    private lateinit var mTurn: ComponentMapper<TurnComponent>
    private lateinit var mUnit: ComponentMapper<UnitComponent>
    private lateinit var mStats: ComponentMapper<StatsComponent>

    private val unitSubscription by lazy {
        world.aspectSubscriptionManager.get(Aspect.all(UnitComponent::class.java))
    }

    fun start() {
        phase = Phase.START
    }

    fun addFaction(faction: Faction) {
        if (factionList.contains(faction)) return
        factionList.add(faction)
    }

    fun removeFaction(faction: Faction) {
        factionList.remove(faction)
    }

    fun reset() {
        factionList.clear()
        currentFactionIndex = 0
        phase = Phase.NONE
    }

    override fun processSystem() {
        when (phase) {
            Phase.START -> {
                giveTurn(factionList[currentFactionIndex])
                phase = Phase.READY
            }
            Phase.READY -> {
                if (entityIds.isEmpty) {
                    phase = Phase.NEXT_TURN
                }
            }
            Phase.NEXT_TURN -> {
                if (!flowControlSystem.ready) return
                if (currentFactionIndex >= factionList.size - 1) {
                    turn += 1
                    currentFactionIndex = 0
                } else {
                    currentFactionIndex += 1
                }

                val nextFaction = factionList[currentFactionIndex]
                Logger.debug { "Turn ended. Next faction: $nextFaction" }
                giveTurn(nextFaction)
                behaviorSystem.prepare()
                phase = Phase.READY
            }
            else -> {
            }
        }
    }

    fun giveTurn(faction: Faction) {
        val entities = unitSubscription.entities
        val data = entities.data
        for (i in 0 until entities.size()) {
            val unit = data[i]
            val cUnit = mUnit[unit]

            if (cUnit.faction != faction) continue

            cUnit.ap = (cUnit.ap + 4 + mStats[unit].ap).coerceAtMost(Constants.MAX_AP)
            mTurn.create(unit)
        }
    }

    fun endTurn(unit: Int) {
        if (!mTurn.has(unit)) return
        mTurn.remove(unit)
    }
}
